Group text lines if they are centered (#382)

jstockwin opened the pull request:
**Description**

Add a check when checking if lines should be merged to see if they are centered above each other.

Fixes #382

Note: it might also be worth having a read of #383, as there is perhaps a discussion there about how to deal with these checks for a negative line margin...

**How Has This Been Tested?**

[Example PDF](https://www.dropbox.com/s/iesuryj5ter1bax/centered_text_example.pdf?dl=0)

Here is my test code (which simply loads the pdfs and prints the elements):
```
from pdfminer import converter, pdfdocument, pdfinterp, pdfpage, pdfparser
from pdfminer.layout import LTTextContainer, LAParams
path_to_file = "centered_text_example.pdf"

with open(path_to_file, "rb") as pdf_file:
    parser = pdfparser.PDFParser(pdf_file)
    document = pdfdocument.PDFDocument(parser)
    resource_manager = pdfinterp.PDFResourceManager()
    device = converter.PDFPageAggregator(resource_manager, laparams=LAParams())
    interpreter = pdfinterp.PDFPageInterpreter(resource_manager, device)
    for page in pdfpage.PDFPage.create_pages(document):
        interpreter.process_page(page)
        results = device.get_result()
        elements = [
            element for element in results if isinstance(element, LTTextContainer)
        ]
device.close()

print(elements)
```

Before this change, you get:
```
[
    <LTTextBoxHorizontal(0) 123.800,344.802,216.392,367.878 'Long Text 1\n'>, 
    <LTTextBoxHorizontal(1) 146.100,366.002,194.070,411.778 'Text 1\nText 2\n'>, 
    <LTTextBoxHorizontal(2) 146.100,302.502,194.070,348.278 'Text 3\nText 4\n'>, 
    <LTTextBoxHorizontal(3) 39.700,24.176,43.200,42.152 ' \n'>, 
    <LTTextBoxHorizontal(4) 395.500,24.176,399.000,42.152 ' \n'>
]
```
After this change you get:
```
[
    <LTTextBoxHorizontal(0) 123.800,302.502,216.392,411.778 'Text 1\nText 2\nLong Text 1\nText 3\nText 4\n'>,
    <LTTextBoxHorizontal(1) 39.700,24.176,43.200,42.152 ' \n'>
    <LTTextBoxHorizontal(2) 395.500,24.176,399.000,42.152 ' \n'>
]
```

(The last two elements in each are just artefacts in the example pdf, ignore them).

**Checklist**

_I will do these things once I get a bit of guidance on whether this is an acceptable change. I'll also need a pointer on how to add a relevant test._

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the [README.md](../README.md) and other documentation, or I am sure that this is not necessary
- [x] I have added a consice human-readable description of the change to [CHANGELOG.md](../CHANGELOG.md)
- [x] I have added docstrings to newly created methods and classes
- [x] I have optimized the code at least one time after creating the initial version

pietermarsman commented:
> I will do these things once I get a bit of guidance on whether this is an acceptable change.

It is

> I'll also need a pointer on how to add a relevant test.

Look at `tests/test_layout.py` for inspiration.

pietermarsman commented on pdfminer/layout.py:
```diff
@@ -422,7 +422,8 @@ def find_neighbors(self, plane, ratio):
                     (abs(obj.x0-self.x0) < d or
-                     abs(obj.x1-self.x1) < d))]
+                     abs(obj.x1-self.x1) < d or
+                     abs((obj.x0+obj.x1)/2-(self.x0+self.x1)/2) < d))]
```
It always takes me a few minutes to grasp the meaning of all these comparisons. Could you introduce well-named (private) methods on `LTTextLineHorizontal` for these comparisons?

jstockwin commented:
@pietermarsman I've added 4 new commits, addressing your comments and then adding tests.

I couldn't find any tests for the function I've changed, but think what I added tests it fairly well (even if it takes a while of figuring out bounding boxes to work out what it's doing....).

Before my change, this test would fail because the `centrally_aligned_overlapping` text boxes would not be included.

Please could you take a look?

pietermarsman commented on CHANGELOG.md:
```diff
@@ -41,18 +44,18 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 ### Deprecated
 - The argument `_py2_no_more_posargs` because Python2 is removed on January
-, 2020 ([#328](https://github.com/pdfminer/pdfminer.six/pull/328) and 
+, 2020 ([#328](https://github.com/pdfminer/pdfminer.six/pull/328) and
```
Did you change these lines on purpose?

jstockwin replied:
Nope, that's my editor removing whitespace from the ends of lines... will edit that commit.

pietermarsman commented on pdfminer/layout.py:
```diff
@@ -416,13 +416,43 @@ def add(self, obj):
+                     self.__is_right_aligned_with(obj, tolerance=d) or
+                     self.__is_centrally_aligned_with(obj, tolerance=d)))]
+
+    def __is_left_aligned_with(self, other, tolerance=0):
```
A single underscore is enough to signal that this method is private.

jstockwin replied:
Done in new commit "Cosmetic changes from code review"

pietermarsman commented on pdfminer/layout.py:
```diff
@@ -416,13 +416,43 @@ def add(self, obj):
+        """
+        Whether the horizontal center of `other` is within `tolerance`.
+        """
+        return abs((other.x0+other.x1)/2 - (self.x0+self.x1)/2) <= tolerance
```
Add some spaces around the `+` and `/` signs

jstockwin replied:
Done in new commit "Cosmetic changes from code review"

jstockwin commented:
@pietermarsman PR updated with your comments
